File: apps/android/app/src/main/java/fi/eport/searchassistant/util/ColorHex.kt
package fi.eport.searchassistant.util

import androidx.compose.ui.graphics.Color

/// Parses "#rrggbb" or "#rrggbbaa". Falls back to gray on bad input —
/// same fallback shape as iOS's `Color(hex:)` extension.
fun String.toComposeColor(): Color {
    var s = this
    if (s.startsWith("#")) s = s.drop(1)
    if (s.length != 6 && s.length != 8) return Color.Gray
    val value = s.toULongOrNull(radix = 16) ?: return Color.Gray
    return if (s.length == 6) {
        Color(
            red = ((value shr 16) and 0xFFu).toInt() / 255f,
            green = ((value shr 8) and 0xFFu).toInt() / 255f,
            blue = (value and 0xFFu).toInt() / 255f,
        )
    } else {
        Color(
            red = ((value shr 24) and 0xFFu).toInt() / 255f,
            green = ((value shr 16) and 0xFFu).toInt() / 255f,
            blue = ((value shr 8) and 0xFFu).toInt() / 255f,
            alpha = (value and 0xFFu).toInt() / 255f,
        )
    }
}

File: apps/ios/Shared/ColorHex.swift
import SwiftUI

extension Color {
    /// Parses "#rrggbb" or "#rrggbbaa". Falls back to gray on bad input.
    ///
    /// Lives in `Shared/` because participant colours arrive as hex strings
    /// from the API and both the phone and the watch have to draw them.
    init(hex: String) {
        var s = hex
        if s.hasPrefix("#") { s.removeFirst() }
        guard s.count == 6 || s.count == 8,
              let value = UInt64(s, radix: 16) else {
            self = .gray; return
        }
        let r, g, b, a: Double
        if s.count == 6 {
            r = Double((value >> 16) & 0xFF) / 255
            g = Double((value >> 8)  & 0xFF) / 255
            b = Double( value        & 0xFF) / 255
            a = 1
        } else {
            r = Double((value >> 24) & 0xFF) / 255
            g = Double((value >> 16) & 0xFF) / 255
            b = Double((value >> 8)  & 0xFF) / 255
            a = Double( value        & 0xFF) / 255
        }
        self = Color(.sRGB, red: r, green: g, blue: b, opacity: a)
    }
}
